sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -438,192 +423,171 @@ class GpuDynamicPartitionDataSingleWriter(
+            savedStatus.get.tableCaches.clear()
+            // concat the caches and this `batch`
+            val concat = withResource(subTables) { _ =>
               // append `batch` to sub tables
-              subTables += GpuColumnVector.from(batch)
+              subTables += table
+              withResource(Table.concatenate(subTables: _*)) { concat =>
+                GpuColumnVector.from(concat, outDataTypes)
+              }
             }
-            withResource(Table.concatenate(subTables: _*)) { concat =>
-              GpuColumnVector.from(concat, outDataTypes)
+            // write concat table
+            if (!needSplitBatch(
+              maxRecordsPerFile,
+              currentWriterStatus.recordsInFile,
+              concat.numRows))
+            {
+              statsTrackers.foreach(_.newBatch(concat))
+              currentWriterStatus.recordsInFile += concat.numRows()
+              // the `write` function will close `concat`
+              currentWriterStatus.outputWriter.write(
+                concat,
+                statsTrackers
+              )
+            } else {
+              withResource(concat){ _ =>
+                writeBatch(concat)
+              }
             }
-          }
-          // write concat table
-          if (!needSplitBatch(
-            maxRecordsPerFile,
-            currentWriterStatus.recordsInFile,
-            concat.numRows))
-          {
-            statsTrackers.foreach(_.newBatch(concat))
-            currentWriterStatus.recordsInFile += concat.numRows()
-            currentWriterStatus.outputWriter.write(concat, statsTrackers)
           } else {
-            withResource(concat){concat =>  
-              writeBatch(concat)
-            }
-          }
-        } else {
-          if (!needSplitBatch(
-            maxRecordsPerFile,
-            currentWriterStatus.recordsInFile,
-            batch.numRows))
-          {
-            statsTrackers.foreach(_.newBatch(batch))
-            currentWriterStatus.recordsInFile += batch.numRows()
-            currentWriterStatus.outputWriter.write(batch, statsTrackers)
-          } else {
-            withResource(batch){batch =>  
-              writeBatch(batch)
+            withResource(GpuColumnVector.from(table, outDataTypes)) { batch =>
+              if (!needSplitBatch(
+                maxRecordsPerFile,
+                currentWriterStatus.recordsInFile,
+                batch.numRows))
+              {
+                statsTrackers.foreach(_.newBatch(batch))
+                currentWriterStatus.recordsInFile += batch.numRows()
+                // increase the reference count of `batch` here because
+                // the `write` function will close `batch`.
+                currentWriterStatus.outputWriter.write(
+                  GpuColumnVector.incRefCounts(batch),
+                  statsTrackers
+                )
+              } else {
+                writeBatch(batch)
+              }
             }
```

Comment:
I still think the batch is being created too early.  If the write implementation can be clever about freeing memory early as the write is being processed, that is defeated by this code which artificially forces the batch to survive through the entire write.  Here's a way to avoid the increfcount hack:
```suggestion
            if (!needSplitBatch(
              maxRecordsPerFile,
              currentWriterStatus.recordsInFile,
              table.getRowCount.toInt))
            {
              val batch = GpuColumnVector.from(table, outDataTypes))
              closeOnExcept(batch) { _ =>
                statsTrackers.foreach(_.newBatch(batch))
                currentWriterStatus.recordsInFile += batch.numRows()
              }
              currentWriterStatus.outputWriter.write(batch, statsTrackers)
            } else {
              withResource(GpuColumnVector.from(table, outDataTypes)) { batch =>
                writeBatch(batch)
              }
            }
          }
```